core/src/main/scala/kafka/server/DelayedFetch.scala

```diff
@@ -148,8 +148,8 @@ class DelayedFetch(delayMs: Long,
             debug(s"Partition $topicIdPartition is in an offline log directory, satisfy $fetchMetadata immediately")
             return forceComplete()
           case _: FencedLeaderEpochException => // Case E
-            debug(s"Broker is the leader of partition $topicIdPartition, but the requested epoch " +
-              s"$fetchLeaderEpoch is fenced by the latest leader epoch, satisfy $fetchMetadata immediately")
+            debug(s"The requested epoch $fetchLeaderEpoch of partition $topicIdPartition" +
```

Comment:
Just for my own understanding -- can the follower get the FenchedLeaderEpochException still?

Reply:
@jolshan Thanks for your review, I read this part of the code again and I think this case exists.
For example:
A partition P0 (1,2,3) with three replicas. 1 is the leader, 2,3 are the followers, the leader epoch is 10, and the consumer fetches data from follower replica 3.
At this time, the leader changes, 2 becomes the leader, the leader epoch becomes 11,  Follower replica 3 receives LeaderAndIsrRequest and executes the makeFollower operation and  updates the leaderEpoch to 11.
If the consumer still uses leader epoch 10 to fetch data from replica 3. it will throw FenchedLeaderEpochException.

Reply:
Interesting -- I was curious since fetch from follower by definition doesn't need the broker to be the leader (so in theory, it could still fetch despite the leader change), though I see how the epoch is necessary to ensure the metadata isn't stale about other aspects too? The error message just seems a little funny given fetch from follower. :)

Reply:
@jolshan I understand one possible reason is that the server's check of the leader epoch is consistent with the client's implementation.
The consumer sends a request to the leader replica, and the leader selects the preferred read replica (possibly a follower replica) and returns it to the consumer.The consumer then fetches data from the preferred read replica.
When the following cases occur, the consumer will return to fetch data from the leader
case1: preferred read replica has expired
case2: received out of range exception from follower replica
case3: It is found that the latest leader epoch of the partition has changed (usually from metadata).
When the consumer fetches data from the Follower replica. I think the epoch check on the server side is consistent with case3.